spec/saml_authenticator_spec.rb

```diff
@@ -425,20 +439,114 @@ def auth_hash(attributes)
 
           result = authenticator.after_authenticate(hash)
 
           expect(result.user.groups.pluck(:name)).to contain_exactly(group1.name, group2.name)
         end
       end
+
+      describe "saml_groups_attribute" do
+        it "syncs groups from the saml_groups_attribute setting" do
+          SiteSetting.saml_groups_attribute = "notTheDefault"
+          hash = auth_hash("notTheDefault" => [group1.name, group2.name])
+
+          result = authenticator.after_authenticate(hash)
+          expect(result.user.groups.pluck(:name)).to contain_exactly(
+            original_group.name,
+            group1.name,
+            group2.name,
+          )
+        end
+
+        it "removes groups from the previously saved saml_groups_attributes in raw_info" do
+          SiteSetting.saml_groups_attribute = "notTheDefault"
+
+          # user's existing group associations and user_associated_account
+          group1.add(user)
+          user.user_associated_accounts.create!(
+            provider_name: "saml",
+            provider_uid: uid,
+            user:,
+            extra: {
+              raw_info: {
+                "memberOf" => [group_without_fullname.name], # this is ignored as it is not the correct attribute
+                "notTheDefault" => [group1.name], # this is the correct attribute
+              },
+            },
+          )
+
+          # new auth hash with a different group
+          hash = auth_hash("notTheDefault" => [group2.name])
+
+          result = authenticator.after_authenticate(hash)
+          expect(result.user.groups.pluck(:name)).to contain_exactly(
+            # group1 should be removed
+            original_group.name,
+            group2.name,
+          )
+        end
+
+        it "functions correctly even with no prior user_associated_account" do
+        end
+
+        it "allows the attribute to specify an array, and assigns groups from those attributes" do
+          SiteSetting.saml_groups_attribute = "Country|Hemisphere"
+          hash =
+            auth_hash(
+              "Country" => [group1.name, group2.name],
+              "Hemisphere" => [group_without_fullname.name],
+            )
+
+          result = authenticator.after_authenticate(hash)
+          expect(result.user.groups.pluck(:name)).to contain_exactly(
+            original_group.name,
+            group1.name,
+            group2.name,
+            group_without_fullname.name,
+          )
+        end
+      end
+
+      describe "saml_groups_use_full_name" do
+        before { SiteSetting.saml_groups_use_full_name = true }
+
+        it "adds users to groups based on group's case insensitive full_names" do
+          SiteSetting.saml_groups_attribute = "oneAttribute|twoAttribute" # ensure compat
+          SiteSetting.saml_sync_groups_list = [group1.full_name, group2.full_name].join("|") # ensure compat
+
+          hash =
+            auth_hash(
+              "oneAttribute" => [group1.full_name, "I don't exist"],
```

Comment:
```suggestion
              "oneAttribute" => [group1.full_name.upcase, "I don't exist"],
```

This will help to test for case insensitivity.